doc/SAI-Proposal-Generic-Counters.md

```diff
@@ -0,0 +1,113 @@
+# Generic SAI Counters and Resources
+
+Title       | Critical Resource Monitor
+------------|----------------
+Authors     | Mellanox
+Status      | In review
+Type        | Standards track
+Created     | 04/10/2019
+SAI-Version | 1.4
+
+## Overview
+ASICs usually provide a pool of generic counters that can be attached to dynamic entries for a better visibility and debuggability.
+This document defines an API for the generic counter object type, that can be attached to one or more objects or entries for reading packet hits on them.
+
+## Counter object type
+For the counter object, a sai_counter_stat_t enum is defined:
+```
+typedef enum _sai_counter_stat_t
+{
+    /** Get tx packets count [uint64_t] */
+    SAI_COUNTER_STAT_PACKETS = 0x00000000,
+
+    /** Get tx bytes count [uint64_t] */
+    SAI_COUNTER_STAT_BYTES = 0x00000001,
+
+     /** Custom range base value */
+    SAI_COUNTER_STAT_CUSTOM_RANGE_BASE = 0x10000000
+
+} sai_counter_stat_t
+```
+
+### Attaching counter to object
+```
+/**
+ * @brief Attribute Id in sai_set_port_attribute() and
+ * sai_get_port_attribute() calls
+ */
+typedef enum _sai_route_entry_attr_t
+{
+    …
+         /**
+     * @brief Attach a list of counters
+     *
+     * When it is empty, then packet hits won't be counted
+     *
+     * @type sai_object_list_t
+     * @flags CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_COUNTER
+     * @default empty
+     */
+    SAI_ROUTE_ENTRY_ATTR_COUNTER_ID_LIST,
```

Comment:
I see a list here and not in the example - whats the final choice ?

Reply:
The final choice is to start with a single id - it's simpler and object query can be implemented with today's API